### ISSUE TITLE
Naming conflict in callback typing produces unnecessry complex state machine

### ISSUE THREAD

**Issue by ulrichb** [babel/babel#11344]
## Bug Report

- [ ] I would like to work on a fix!

**Current Behavior**
A clear and concise description of the behavior.

**Input Code**
- REPL or Repo link if applicable:

```ts
async function someAsync(
    factory: () => Promise<string>,
    options: {
        someCallback?: (theSameName: number) => void;
    } = {},
): Promise<string> {

    const theSameName = await factory();

    return theSameName;
}
```

[REPL Link](https://babeljs.io/en/repl#?browsers=ie%2011&build=&builtIns=false&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsIxOgtgUwEFxoAKAKBivmCmXQCcwAuGUgShgF4A-GABQZ5UIfAB4QyBqggBzHgBpK1dAAc0mEKwDey6tRwEAwsAA2pgEa0A1gH5WpZAAt8AZWAEAch_ysICXAt8Bk5eGAA3dFQAEwBuPRgAX24YbUSldlZBYVEJKRl5VPIEqE1kGGc3H28CFOAAd2BUcrhaeiYOeISGfGQEBixK9y8feMSgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env%2Ctypescript&prettier=false&targets=&version=7.9.0&externalPlugins=)

**Expected behavior/code**

When the argument of the `someCallback` _typing_ has the same name as the local in the body, then the generated state machine becomes weird (two nested `regeneratorRuntime.wrap()`). See the following text comparison of the REPL output:

![image](https://user-images.githubusercontent.com/388796/77775227-90f4a900-704c-11ea-9474-1b3d21b6faad.png)

Expected: No difference when the argument name in the callback _typing_ is the same as a local.

**Babel Configuration (babel.config.js, .babelrc, package.json#babel, cli command, .eslintrc)**

Just the TypeScript preset, see REPL link above.

**Additional info**

This is a follow-up issue of https://github.com/babel/babel/issues/11336#issuecomment-605075901, where I get _"Module parse failed: Cannot use keyword 'await' outside an async function"_ errors in the CRA Babel preset.

**Comment by nicolo-ribaudo** [comment]
Simpler example:
```js
function* a(b: (c) => void = {}) {
  const c = yield 0;
  return c;
}
```

**Comment by nicolo-ribaudo** [comment]
@ulrichb Can you check the version of `@babel/plugin-transform-parameters`? This seems to be fixed in 7.9.3: https://babeljs.io/en/repl#?browsers=ie%2011&build=&builtIns=false&spec=false&loose=false&code_lz=IYZwngdgxgBAZgV2gFwJYHsIxOgtgUwEFxoAKAKBivmCmXQCcwAuGUgShgF4A-GABQZ5UIfAB4QyBqggBzHgBpK1dAAc0mEKwDey6tRwEAwsAA2pgEa0A1gH5WpZAAt8AZWAEAch_ysICXAt8Bk5eGAA3dFQAEwBuPRgAX24YbUSldlZBYVEJKRl5VPIEqE1kGGc3H28CFOAAd2BUcrhaeiYOeISGfGQEBixK9y8feMSgA&debug=false&forceAllTransforms=false&shippedProposals=false&circleciRepo=&evaluate=false&fileSize=false&timeTravel=false&sourceType=module&lineWrap=false&presets=env%2Ctypescript%2Cenv&prettier=false&targets=&version=7.9.3&externalPlugins= (the repl will show 7.9.0 because it shows `@babel/core`'s version).

**Comment by ulrichb** [comment]
@nicolo-ribaudo Okay, cool. Thanks!
